CMakeLists.txt

```diff
@@ -1,330 +1,712 @@
+
+# provide user options
+option (BUILD_SHARED_LIBS       "Build NLopt as a shared library" ON)
+option (WITH_CXX                "Enable C++-based routines (StoGO)" OFF)
+option (WITH_SWIG               "Use SWIG to generate bindings" ON)
+option (BUILD_CPP               "Build C++ interace (nlopt.hpp)" ON)
+option (BUILD_F77               "Build Fortran iterface" ON)
+option (BUILD_PYTHON            "Build Python bindings (using SWIG)" ON)
+option (BUILD_GUILE             "Build Guile bindings (using SWIG)" ON)
+option (BUILD_OCTAVE            "Build Octave bindings" ON)
+option (BUILD_MATLAB            "Build MATLAB bindings" ON)
+option (BUILD_TESTS             "Build tests" ON)
+option (BUILD_TESTS_UTIL        "Build util tests" OFF)
+option (BUILD_TESTS_ALG         "Build algorithms tests" OFF)
+option (WITH_TESTS_INSTALL_TREE "Run tests from install-tree or build-tree" OFF)
+option (WITH_EXTRA_WARNINGS     "Enable extra compiler warnings (GNU GCC)" OFF)
+option (WITH_FOLDERS            "Enable grouping of files/targets in folders" ON)
+
+# validate options combination
+if (BUILD_PYTHON OR BUILD_GUILE)
+  if (NOT WITH_SWIG)
+    message (WARNING "Python/Guile wrappers require SWIG")
+  endif ()
+  if (NOT BUILD_CPP)
+    message (WARNING "Python/Guile wrappers require the C++ interface (nlopt.hpp)")
+  endif ()
+endif ()
+
+# enable C++ language compiler if needed
+if (WITH_CXX OR BUILD_PYTHON OR BUILD_GUILE OR BUILD_OCTAVE)
+  enable_language (CXX)
+endif ()
+
+# default installation directories
+# (provides variables: CMAKE_INSTALL_<dir> and CMAKE_INSTALL_FULL_<dir>)
+include (GNUInstallDirs)
+
+# set RPATH of installed executables and libraries
+set (CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_FULL_LIBDIR}")
+set (CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
+if (POLICY CMP0042)
+  # Set MACOSX_RPATH to ON
+  cmake_policy (SET CMP0042 NEW)
+endif ()
+
+# organize target projects inside IDE
+# (group cmake generated targets like ALL_BUILD/ZERO_CHECK/..)
+if (WITH_FOLDERS)
+  set_property (GLOBAL PROPERTY USE_FOLDERS ON)
+endif ()
+
+# set compiler-specific flags
+if (CMAKE_C_COMPILER_ID STREQUAL "GNU" OR CMAKE_C_COMPILER_ID STREQUAL "Clang")
+  if (WITH_EXTRA_WARNINGS)
+    # add extra warnings in maintainer mode
+    foreach (_FLG -Wall -W -Wcast-qual -Wpointer-arith -Wcast-align
+        -Wno-long-long -pedantic -Wshadow -Wwrite-strings -Wredundant-decls
+        -Wundef -Wconversion)
+      set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${_FLG}")
+      set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${_FLG}")
+    endforeach ()
+    foreach (_FLG -Wbad-function-cast -Wstrict-prototypes -Wnested-externs
+        -Wmissing-prototypes -Wmissing-declarations)
+      set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${_FLG}")
+    endforeach ()
+    unset (_FLG)
+  else ()
+    # turn off deprecated warnings
+    set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-deprecated-declarations")
+    set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated-declarations")
+  endif ()
+endif ()
+
```

Comment:
It's simpler to let each developer override its own flags. Wshadow is buggy on gcc anyway.